File: mpi/mpi2Examples/worker.c
#include <stdio.h>
#include "mpi.h"
#include <math.h>
 
void lam_darwin_malloc_linker_hack(){};
void timestmp(char *timestr);

int main(argc,argv)
int argc;
char *argv[];
{
    int myid, numprocs;
    FILE *f1;
    int i;
    int ierr;
    MPI_Comm  parent;
    char fname[10];
    char stamp[11];
 
    MPI_Init(&argc,&argv);
    MPI_Comm_size(MPI_COMM_WORLD,&numprocs);
    MPI_Comm_rank(MPI_COMM_WORLD,&myid);
    MPI_Comm_get_parent(&parent);
 
    sprintf(fname,"%s%2.2d",argv[1],myid);

    printf("Hello from c worker %d writing to %s\n",myid,fname);
    f1=fopen(fname,"w");
    timestmp(stamp);
    fprintf(f1,"%s\n",stamp);
    fclose(f1);
    sleep(10);
    MPI_Comm_free(&parent);
    MPI_Finalize();
    
}


#include <time.h>
#include <sys/time.h>
#include <string.h>
void timestmp(char *timestr) {
/*	char timestr[11]; */
	struct timeval tp;
	struct timezone tzp;
	struct tm *tmstruct;
	time_t tod;
	gettimeofday(&tp,&tzp);
	tod=tp.tv_sec;
	tmstruct=gmtime(&tod);
/*	tmstruct=localtime(&tod); */
	sprintf(timestr,"%2.2d%2.2d%2.2d%2.2d%2.2d",tmstruct->tm_mon+1,
	                                            tmstruct->tm_mday,
	                                            tmstruct->tm_hour,
	                                            tmstruct->tm_min,
	                                            tmstruct->tm_sec);
	timestr[10]=(char)0;
}

File: mpi/mpi2Examples/worker.py
#!/usr/bin/env python
import numpy
from numpy import *
import mpi
import sys
from time import gmtime, time,sleep


def stamp():
	timeTuple = gmtime(time())[1:6]
	return "%02d%02d%02d%02d%02d" % timeTuple


sys.argv =  mpi.mpi_init(len(sys.argv),sys.argv)
myid=mpi.mpi_comm_rank(mpi.MPI_COMM_WORLD)
numprocs=mpi.mpi_comm_size(mpi.MPI_COMM_WORLD)
parent=mpi.mpi_comm_get_parent()
parentSize=mpi.mpi_comm_size(parent)
print "parentSize",parentSize

tod=stamp()
s=sys.argv[1]+"%2.2d" % myid
print "hello from python worker",myid," writing to ",s

x=array([5,3,4,2],'i')
print "starting bcast"
buffer=mpi.mpi_bcast(x,4,mpi.MPI_INT,0,parent)
out = open(s, "w")
out.write(str(buffer))
out.write(tod+"\n")
out.close()

print myid," got ",buffer
junk=mpi.mpi_scatter(x,1,mpi.MPI_INT,1,mpi.MPI_INT,0,parent)
print myid," got scatter ",junk


back=mpi.mpi_recv(1,mpi.MPI_INT,0,1234,parent)
back[0]=back[0]+1
mpi.mpi_send(back,1,mpi.MPI_INT,0,5678,parent)

dummy=myid
final=mpi.mpi_reduce(dummy,1,mpi.MPI_INT,mpi.MPI_SUM,0,parent)

sleep(10)
mpi.mpi_comm_free(parent)
mpi.mpi_finalize()
